### PR TITLE
[Workflow] Add a profiler

### PR DESCRIPTION
**Pull request by lyrixx**
| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       |
| License       | MIT
| Doc PR        |

---

With [this project lyrixx/SFLive-Paris2016-Workflow](https://github.com/lyrixx/SFLive-Paris2016-Workflow)

image

[workflow.webm](https://github.com/symfony/symfony/assets/408368/e86aad22-812d-461d-950f-68d50ac31390)

### CODE REVIEW

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Icon/workflow.svg
Comment (lyrixx) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Icon/workflow.svg:1]: I'm totally a bad UI designer, so I found this on the web, and I adapted it a bit.
Feel free to replace it with something better

cc @jonathanlevaillant

File: src/Symfony/Component/Workflow/DataCollector/WorkflowDataCollector.php [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    public function lateCollect()
+    {
+        foreach ($this->workflows as $workflow) {
+            $type = $workflow instanceof StateMachine ? MermaidDumper::TRANSITION_TYPE_STATEMACHINE : MermaidDumper::TRANSITION_TYPE_WORKFLOW;
+            $dumper = new MermaidDumper($type);
+            $this->data['dumps'][$workflow->getName()] = $dumper->dump($workflow->getDefinition());
+        }
+    }
+
+    public function getName()
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Workflow/DataCollector/WorkflowDataCollector.php]:
```suggestion
    public function getName(): string
```

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/workflow.html.twig [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            // "good size" if they are not visible
+            document.addEventListener('DOMContentLoaded', () => {
+                document.querySelectorAll('.tab').forEach((el) => {
+                    const observer = new MutationObserver(() => {
+                        if (!el.classList.contains('block')) {
+                            return;
+                        }
+                        const mEl = el.querySelector('.sf-mermaid');
+                        if (mEl.dataset.processed) {
+                            console.log('A');
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/workflow.html.twig]: ?

File: src/Symfony/Component/Workflow/DataCollector/WorkflowDataCollector.php (hunk 1 above)
Comment (lyrixx) [reply src/Symfony/Component/Workflow/DataCollector/WorkflowDataCollector.php]: done, thx (not pushed yet).

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/workflow.html.twig (hunk 2 above)
Comment (lyrixx) [reply src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/workflow.html.twig]: fixed

**Review by nicolas-grekas — APPROVED** [review]

**Comment by stof** [comment]
This panel does not collect any info related to the request. It only collect info about the configuration of the project, which will be the same for all requests. In the past, we removed some panels like that.

**Comment by lyrixx** [comment]
I thought about doing everything at runtime (I mean without all the collector stuff). But I changed my mind because the workflow can evolve between 2 requests. And we collect already some information about listener, doctrine entities. API-P does the same with the resources, and so on

And this panel is **really** useful. I'm working on a new projet, with a huge workflow, and it helped me a lot to build the configuration.

And I plan to add more information to it, like the list of all event listeners. This features has been requested in #50976 

So I would **be very happy** to get this merged, it'll be a game changer for everyone using the workflow component

**Comment by fabpot** [comment]
We already one such panel: "configuration"

**Comment by lyrixx** [comment]
Note: In future version, this panal will also display more things

* The current marking of the subject (if a call has been made to the workflow obliviously)
* Current content of `TransitionBlockerList`
* Perf metrics about all listeners

---

So it could not fit in the configuration panel

It will be something like the security panel : some configuration + some request related content

**Comment by 94noni** [comment]
i also find it useful even if not linked to request/response flow
reminds me https://github.com/symfony/symfony/issues/39114

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @lyrixx.

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Icon/workflow.svg
Comment (94noni) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Icon/workflow.svg:1]: perhaps use https://tabler-icons.io/i/vector @lyrixx 
new design profiler take icons from this lib :)

Comment (stof) [reply src/Symfony/Bundle/WebProfilerBundle/Resources/views/Icon/workflow.svg:1]: I would say than https://tabler-icons.io/i/git-fork better represents the idea of workflows than this square that you suggested.

Comment (94noni) [reply src/Symfony/Bundle/WebProfilerBundle/Resources/views/Icon/workflow.svg:1]: indeed! just passing by ;)

Comment (lyrixx) [reply src/Symfony/Bundle/WebProfilerBundle/Resources/views/Icon/workflow.svg:1]: good idea, I'll do that in another PR

**Comment by andreybolonin** [comment]
@lyrixx I used your WorkflowExtension to dump Workflow builded on the fly (in Controller), but I see some blocks is rendered on grey background

Screenshot 2023-10-24 at 13 03 53

How you fixed this?

I used the same code on frontend (mermaid.js and the same ver of bootstrap.min.css)

**Comment by lyrixx** [comment]
I'm not sure to understand your issue. Could could open a github discussion to talk about that? Thanks

**Pull request merged**